src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Ascii.cs

```diff
@@ -0,0 +1,1244 @@
+            }
+
+            if (source.Overlaps(destination))
+            {
+                throw new InvalidOperationException(SR.InvalidOperation_SpanOverlappedOperation);
+            }
+
+            for (int i = 0; i < source.Length; i++)
+            {
+                destination[i] = ToUpper(source[i]);
+            }
+
+            return source.Length;
+        }
+
+        /// <summary>
+        /// Copies the contents of <paramref name="source"/> to <paramref name="destination"/>,
+        /// converting any lowercase ASCII <see cref="char"/>s to their uppercase equivalents
+        /// during the copy.
+        /// </summary>
+        /// <param name="source">The source buffer.</param>
+        /// <param name="destination">The destination buffer.</param>
+        /// <returns>
+        /// The number of <see cref="char"/>s copied to <paramref name="destination"/>.
+        /// This will be equivalent to <paramref name="source"/>'s <see cref="Span{Char}.Length"/>.
+        /// </returns>
+        /// <remarks>
+        /// If <paramref name="source"/> contains non-ASCII data, use <see cref="MemoryExtensions.ToUpperInvariant"/> instead.
+        /// </remarks>
+        /// <exception cref="ArgumentException"><paramref name="destination"/> is not large enough
+        /// to contain the copied data.</exception>
+        /// <exception cref="InvalidOperationException"><paramref name="source"/> and <paramref name="destination"/> overlap.</exception>
+        public static int ToUpper(ReadOnlySpan<char> source, Span<char> destination)
+        {
+            if (destination.Length < source.Length)
+            {
+                ThrowHelper.ThrowArgumentException_DestinationTooShort();
+            }
+
+            if (source.Overlaps(destination))
+            {
+                throw new InvalidOperationException(SR.InvalidOperation_SpanOverlappedOperation);
+            }
+
+            for (int i = 0; i < source.Length; i++)
+            {
+                destination[i] = ToUpper(source[i]);
+            }
+
+            return source.Length;
+        }
+
+        /*
+         * Performs case conversion ([A-Z] -> [a-z] or vice versa) in-place. All values
+         * outside [A-Za-z] - including non-ASCII values - are unchanged.
+         */
+
+        /// <summary>
+        /// Performs an in-place lowercase ASCII conversion of a buffer's contents.
+        /// </summary>
+        /// <param name="value">The buffer over which to operate.</param>
+        /// <remarks>
+        /// The contents of <paramref name="value"/> are modified in-place. Uppercase
+        /// ASCII <see cref="byte"/>s are converted to lowercase. Non-ASCII <see cref="byte"/>s
+        /// remain unchanged.
+        /// </remarks>
+        public static void ToLowerInPlace(Span<byte> value)
+        {
+            foreach (ref byte refElement in value)
+            {
+                refElement = ToLower(refElement);
+            }
+        }
```

Comment:
Clever.

Reply:
With `refElement = ToLower(ref refElement)` the JIT produces even better code (assuming ToLower is changed accordingly to take `ref`):
```diff
- movzx r9d, byte [r8]
- or r9d, 0x20
- mov [r8], r9b
+ or byte [r8], 0x20
```

Reply:
Writes are somewhat expensive. The ideal implementation of this method wouldn't write back to memory unless the data needed to be converted.

But that should be easy enough to add as part of a "read more than one byte at a time" pass. We can read 4 or 8 bytes at a time even without involving SIMD.

Reply:
Out of curiosity:

> ideal implementation of this method wouldn't write back to memory unless the data needed to be converted

For single `byte`:
Do you have any idea how to accomplish this without introducing (non predictable) branches or using `cmov` (which may harm speculations in the cpu-pipeline)?
I think the best compromise is just to write back the data to memory as it's done right now.

For multiple `byte`s with SWAR / SIMD this cost should be amortized -- as you pointed out.